src/lsm/grid.zig

```diff
@@ -381,119 +357,99 @@ pub fn GridType(comptime Storage: type) type {
+                while (it) |queued_read| : (it = queued_read.next) {
+                    if (address == queued_read.address) {
+                        assert(checksum == queued_read.checksum);
+                        assert(block_type == queued_read.block_type);
+                        queued_read.resolves.push(&read.pending);
                         return;
                     }
                 }
             }
 
-            // If the block is already in the cache, queue up the read to be resolved
-            // from the cache on the next tick. This keeps start_read() asynchronous.
-            // Note that this must be called after we have checked for an in
-            // progress read targeting the same address, otherwise we may read an
-            // unininitialized block.
-            if (grid.cache.exists(read.address)) {
-                grid.read_cached_queue.push(read);
+            // Become the "root" read thats fetching the block for the given address.
+            // The fetch happens asynchronously to avoid stack-overflow and nested cache invalidation.
+            grid.read_queue.push(read);
+            grid.on_next_tick(read_block_tick_callback, &read.next_tick);
+        }
+
+        fn read_block_tick_callback(next_tick: *Storage.NextTick) void {
+            const read = @fieldParentPtr(Grid.Read, "next_tick", next_tick);
+            const grid = read.grid;
+
+            // Try to resolve the read from the cache.
+            if (grid.cache.get(read.address)) |block| {
+                if (constants.verify) grid.verify_cached_read(read.address, block);
+                grid.read_block_resolve(read, block);
                 return;
             }
 
+            // Grab an IOP to resolve the block from storage.
+            // Failure to do so means the read is queued to receive an IOP when one finishes.
             const iop = grid.read_iops.acquire() orelse {
-                grid.read_queue.push(read);
+                grid.read_pending_queue.push(&read.pending);
                 return;
             };
 
+            grid.read_block_with(iop, read);
+        }
+
+        fn read_block_with(grid: *Grid, iop: *Grid.ReadIOP, read: *Grid.Read) void {
+            const address = read.address;
+            assert(address > 0);
+
+            // We can only update the cache if the Grid is not resolving callbacks with a cache block.
+            assert(!grid.read_resolving);
+
+            // Grab a block from the cache to read with.
+            // This also inserts the block into the cache at the address.
             const block = grid.cache.insert_preserve_locked(
                 *Grid,
                 block_locked,
                 grid,
-                read.address,
+                address,
             );
+
             // `block` will be initialized later when the read completes.
             // This is safe because we will never attempt to lookup an address in the cache
```

Comment:
Let's make this reasoning more explicit.

```suggestion
            // This is safe because as long as `read` is in `grid.read_queue` or `grid.read_recovery_queue` we will never attempt to read from or overwrite this cache entry.
```